# **Salifort Motors: Providing Insights and Data Driven Suggestions for HR**

## **Pace Stage 1: Plan**

### Understand the business scenario and problem:

The HR department at Salifort Motors wants to take some initiatives to improve employee satisfaction levels at the company. They collected data from employees, but now they don’t know what to do with it. They refer to me as a data analytics professional and asked me to provide data-driven suggestions based on my understanding of the data. 

**They have the following question:** What’s likely to make an employee leave the company?

My goals in this project are to analyze the data collected by the HR department and to build a model that predicts whether or not an employee will leave the company. 

If I can predict employees likely to quit, it might be possible to identify factors that contribute to their leaving. Because it is time-consuming and expensive to find, interview, and hire new employees, increasing employee retention will be beneficial to the company.

### Data available:

According to the HR department, in the [HR_dataset.csv](HR_dataset.csv), there are 14,999 rows, 10 columns, and these variables: 

| Variable  |Description |
| ----- | ----- | 
| satisfaction_level | Employee-reported job satisfaction level [0&ndash;1]|
| last_evaluation | Score of employee's last performance review [0&ndash;1]|
| number_project | Number of projects employee contributes to|
| average_monthly_hours | Average number of hours employee worked per month|
| time_spend_company | How long the employee has been with the company (years) |
| Work_accident | Whether or not the employee experienced an accident while at work |
| left | Whether or not the employee left the company |
| promotion_last_5years | Whether or not the employee was promoted in the last 5 years |
| Department | The employee's department |
| salary | The employee's salary (U.S. dollars) |

### **Data Exploration, Cleaning and Analysis (Initial EDA)**

#### Step 1: Import Packages

In [1]:
# Import packages
### YOUR CODE HERE ### 

# For data manipulation
import numpy as np
import pandas as pd

# For data visualization
import matplotlib.pyplot as plt
import seaborn as sns

# For displaying all of the columns in dataframes
pd.set_option('display.max_columns', None)

#### Step 2: Load the Data

In [2]:
# Load dataset into a dataframe
df0 = pd.read_csv("HR_dataset.csv")

# Display first 10 rows of the dataframe
df0.head(10)

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2                   157   
1                0.80             0.86               5                   262   
2                0.11             0.88               7                   272   
3                0.72             0.87               5                   223   
4                0.37             0.52               2                   159   
5                0.41             0.50               2                   153   
6                0.10             0.77               6                   247   
7                0.92             0.85               5                   259   
8                0.89             1.00               5                   224   
9                0.42             0.53               2                   142   

   time_spend_company  Work_accident  left  promotion_last_5years Department  \
0                   3              0     1                      0      sales   
1                   6              0     1                      0      sales   
2                   4              0     1                      0      sales   
3                   5              0     1                      0      sales   
4                   3              0     1                      0      sales   
5                   3              0     1                      0      sales   
6                   4              0     1                      0      sales   
7                   5              0     1                      0      sales   
8                   5              0     1                      0      sales   
9                   3              0     1                      0      sales   

   salary  
0     low  
1  medium  
2  medium  
3     low  
4     low  
5     low  
6     low  
7     low  
8     low  
9     low

**Note:** The information provided by the HR department about the "salary" field is incorrect. The "salary" field is not defined as numerical values (U.S. dollars) but categorical values.

#### Step 3: Gather basic information about the data

In [3]:
# Gather basic information about the data
df0.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   satisfaction_level     14999 non-null  float64
 1   last_evaluation        14999 non-null  float64
 2   number_project         14999 non-null  int64  
 3   average_montly_hours   14999 non-null  int64  
 4   time_spend_company     14999 non-null  int64  
 5   Work_accident          14999 non-null  int64  
 6   left                   14999 non-null  int64  
 7   promotion_last_5years  14999 non-null  int64  
 8   Department             14999 non-null  object 
 9   salary                 14999 non-null  object 
dtypes: float64(2), int64(6), object(2)
memory usage: 1.1+ MB


#### Step 4: Gather basic descriptive statistics about the data

In [4]:
# Gather basic descriptive statistics about the data
df0.describe(include='all')

satisfaction_level  last_evaluation  number_project  \
count         14999.000000     14999.000000    14999.000000   
unique                 NaN              NaN             NaN   
top                    NaN              NaN             NaN   
freq                   NaN              NaN             NaN   
mean              0.612834         0.716102        3.803054   
std               0.248631         0.171169        1.232592   
min               0.090000         0.360000        2.000000   
25%               0.440000         0.560000        3.000000   
50%               0.640000         0.720000        4.000000   
75%               0.820000         0.870000        5.000000   
max               1.000000         1.000000        7.000000   

        average_montly_hours  time_spend_company  Work_accident          left  \
count           14999.000000        14999.000000   14999.000000  14999.000000   
unique                   NaN                 NaN            NaN           NaN   
top                      NaN                 NaN            NaN           NaN   
freq                     NaN                 NaN            NaN           NaN   
mean              201.050337            3.498233       0.144610      0.238083   
std                49.943099            1.460136       0.351719      0.425924   
min                96.000000            2.000000       0.000000      0.000000   
25%               156.000000            3.000000       0.000000      0.000000   
50%               200.000000            3.000000       0.000000      0.000000   
75%               245.000000            4.000000       0.000000      0.000000   
max               310.000000           10.000000       1.000000      1.000000   

        promotion_last_5years Department salary  
count            14999.000000      14999  14999  
unique                    NaN         10      3  
top                       NaN      sales    low  
freq                      NaN       4140   7316  
mean                 0.021268        NaN    NaN  
std                  0.144281        NaN    NaN  
min                  0.000000        NaN    NaN  
25%                  0.000000        NaN    NaN  
50%                  0.000000        NaN    NaN  
75%                  0.000000        NaN    NaN  
max                  1.000000        NaN    NaN

#### Step 5: Data cleaning

Standardize the column names so that they are all in `snake_case` and correct posible spelling mistakes.

In [5]:
# Get all column names
df0.columns

Index(['satisfaction_level', 'last_evaluation', 'number_project',
       'average_montly_hours', 'time_spend_company', 'Work_accident', 'left',
       'promotion_last_5years', 'Department', 'salary'],
      dtype='object')

In [6]:
# Rename column as needed
df0 =df0.rename(columns={ 'Work_accident':'work_accident',
                          'average_montly_hours': 'average_monthly_hours',
                          'time_spend_company': 'tenure',
                          'Department': 'department'
                          })

df0.columns

Index(['satisfaction_level', 'last_evaluation', 'number_project',
       'average_monthly_hours', 'tenure', 'work_accident', 'left',
       'promotion_last_5years', 'department', 'salary'],
      dtype='object')

#### Step 5: Check for missing values

**Note:** From the basic information gathered in step 3, I conclude that there are no missing values in the data, but lets validate that conclusion.

In [7]:
# Check for missing values
df0.isna().sum()

satisfaction_level       0
last_evaluation          0
number_project           0
average_monthly_hours    0
tenure                   0
work_accident            0
left                     0
promotion_last_5years    0
department               0
salary                   0
dtype: int64

**Note:** Confirmed, there are no missing values in the data.

#### Step 6: Check for duplicates

In [8]:
# Check for duplicates
duplicates = df0.duplicated().sum()

if duplicates>0 :
    print('There are', duplicates, 'duplicated rows in the data.')
    total_rows = df0['salary'].count()
    print('Percentage of dupicated data:', round(duplicates/total_rows*100, 2) ,'%')
else:
    print('There are no duplicates in the data.')

There are 3008 duplicated rows in the data.
Percentage of dupicated data: 20.05 %


3008 rows contain duplicates. **That is approximately 20% of the total rows in the data.** Further inspection is needed to determine if this duplicates are legitimate data or not.

In [25]:
# Inspect rows that contain duplicates
df0[df0.duplicated()].head(20)

satisfaction_level  last_evaluation  number_project  \
396                  0.46             0.57               2   
866                  0.41             0.46               2   
1317                 0.37             0.51               2   
1368                 0.41             0.52               2   
1461                 0.42             0.53               2   
1516                 0.40             0.50               2   
1616                 0.37             0.46               2   
1696                 0.39             0.56               2   
1833                 0.10             0.85               6   
12000                0.38             0.53               2   
12001                0.80             0.86               5   
12002                0.11             0.88               7   
12003                0.72             0.87               5   
12004                0.37             0.52               2   
12005                0.41             0.50               2   
12006                0.10             0.77               6   
12007                0.92             0.85               5   
12008                0.89             1.00               5   
12009                0.42             0.53               2   
12010                0.45             0.54               2   

       average_monthly_hours  tenure  work_accident  left  \
396                      139       3              0     1   
866                      128       3              0     1   
1317                     127       3              0     1   
1368                     132       3              0     1   
1461                     142       3              0     1   
1516                     127       3              0     1   
1616                     156       3              0     1   
1696                     160       3              0     1   
1833                     266       4              0     1   
12000                    157       3              0     1   
12001                    262       6              0     1   
12002                    272       4              0     1   
12003                    223       5              0     1   
12004                    159       3              0     1   
12005                    153       3              0     1   
12006                    247       4              0     1   
12007                    259       5              0     1   
12008                    224       5              0     1   
12009                    142       3              0     1   
12010                    135       3              0     1   

       promotion_last_5years  department  salary  
396                        0       sales     low  
866                        0  accounting     low  
1317                       0       sales  medium  
1368                       0       RandD     low  
1461                       0       sales     low  
1516                       0          IT     low  
1616                       0       sales     low  
1696                       0       sales     low  
1833                       0       sales     low  
12000                      0       sales     low  
12001                      0       sales  medium  
12002                      0       sales  medium  
12003                      0       sales     low  
12004                      0       sales     low  
12005                      0       sales     low  
12006                      0       sales     low  
12007                      0       sales     low  
12008                      0       sales     low  
12009                      0       sales     low  
12010                      0       sales     low

In [27]:
# Inspect rows that contain duplicates
df0[df0.duplicated()].tail(20)

satisfaction_level  last_evaluation  number_project  \
14979                0.09             0.93               6   
14980                0.76             0.89               5   
14981                0.73             0.93               5   
14982                0.38             0.49               2   
14983                0.72             0.84               5   
14984                0.40             0.56               2   
14985                0.91             0.99               5   
14986                0.85             0.85               4   
14987                0.90             0.70               5   
14988                0.46             0.55               2   
14989                0.43             0.57               2   
14990                0.89             0.88               5   
14991                0.09             0.81               6   
14992                0.40             0.48               2   
14993                0.76             0.83               6   
14994                0.40             0.57               2   
14995                0.37             0.48               2   
14996                0.37             0.53               2   
14997                0.11             0.96               6   
14998                0.37             0.52               2   

       average_monthly_hours  tenure  work_accident  left  \
14979                    296       4              0     1   
14980                    238       5              0     1   
14981                    162       4              0     1   
14982                    137       3              0     1   
14983                    257       5              0     1   
14984                    148       3              0     1   
14985                    254       5              0     1   
14986                    247       6              0     1   
14987                    206       4              0     1   
14988                    145       3              0     1   
14989                    159       3              1     1   
14990                    228       5              1     1   
14991                    257       4              0     1   
14992                    155       3              0     1   
14993                    293       6              0     1   
14994                    151       3              0     1   
14995                    160       3              0     1   
14996                    143       3              0     1   
14997                    280       4              0     1   
14998                    158       3              0     1   

       promotion_last_5years department  salary  
14979                      0  technical  medium  
14980                      0  technical    high  
14981                      0  technical     low  
14982                      0  technical  medium  
14983                      0  technical  medium  
14984                      0  technical  medium  
14985                      0  technical  medium  
14986                      0  technical     low  
14987                      0  technical     low  
14988                      0  technical     low  
14989                      0  technical     low  
14990                      0    support     low  
14991                      0    support     low  
14992                      0    support     low  
14993                      0    support     low  
14994                      0    support     low  
14995                      0    support     low  
14996                      0    support     low  
14997                      0    support     low  
14998                      0    support     low

Considering that there are no employee_id (anonymous survey) or record_id fields in the data, I can not conclude that these are "true" duplicates. I would recommend to the HR department to include a record_id field for future surveys to facilitate the detection of duplicated data entries.

Taking into account that the first few rows of "duplicated" data appear to aleatory errors and that the probability of two or more people getting the same answers on ten different continous variable fields is very low, I can conclude that those records should be considered "typos" or duplicated entries to the survey.

On the other hand, there is a high number of duplicated entries that goes from the index number 12000 to 14998 in increments of 1. This makes it look like a duplicated data loading.

In [31]:
df0[df0.duplicated()]['left'].value_counts()

left
1    1580
0    1428
Name: count, dtype: int64